### Задание 2

In [1]:
!python -m cProfile -s time dd.py

         39378491 function calls (39378212 primitive calls) in 15.108 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      240    8.280    0.035    8.570    0.036 dd.py:172(dexor_table)
      157    3.453    0.022    5.786    0.037 dd.py:77(read_string)
 36755509    2.280    0.000    2.280    0.000 {built-in method builtins.chr}
    30848    0.258    0.000    0.358    0.000 dd.py:134(unpack_wide_string)
    39726    0.217    0.000    0.217    0.000 {method 'join' of 'str' objects}
  1550827    0.190    0.000    0.190    0.000 {built-in method builtins.ord}
       23    0.121    0.005    8.386    0.365 dd.py:227(export_field)
   175133    0.085    0.000    0.103    0.000 dd.py:85(get_packed_value)
        1    0.063    0.063   15.095   15.095 dd.py:348(decode)
   213942    0.028    0.000    0.028    0.000 {method 'append' of 'list' objects}
285364/285300    0.027    0.000    0.027    0.000 {built-in method builtins.len}
   

Export step 1. First while
Export step 2. Second while
Export step 3. Third while
Export step 4. org
Export step 5. city
Export step 6. street
Export step 7. fil_contact
Export step 8. fil_org
Export step 9. fil_address
Export step 10. address_elem
Export step 11. org_rub
Export step 12. fil_rub
Export step 13. rub3
Export step 14. rub2
Export step 15. rub1


Испрвил - dexor_table

In [2]:
!python -m cProfile -s time dd_optimised.py

         38751775 function calls (38751496 primitive calls) in 7.089 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      157    3.512    0.022    5.898    0.038 dd_optimised.py:77(read_string)
 36128784    2.261    0.000    2.261    0.000 {built-in method builtins.chr}
      240    0.255    0.001    0.440    0.002 dd_optimised.py:172(dexor_table)
    30848    0.250    0.000    0.348    0.000 dd_optimised.py:134(unpack_wide_string)
    39735    0.222    0.000    0.222    0.000 {method 'join' of 'str' objects}
       23    0.127    0.006    1.069    0.046 dd_optimised.py:240(export_field)
   840667    0.099    0.000    0.099    0.000 {method 'append' of 'list' objects}
   175133    0.090    0.000    0.109    0.000 dd_optimised.py:85(get_packed_value)
   924102    0.080    0.000    0.080    0.000 {built-in method builtins.ord}
        1    0.061    0.061    7.076    7.076 dd_optimised.py:361(decode)
285364/285300    0.029   

Export step 1. First while
Export step 2. Second while
Export step 3. Third while
Export step 4. org
Export step 5. city
Export step 6. street
Export step 7. fil_contact
Export step 8. fil_org
Export step 9. fil_address
Export step 10. address_elem
Export step 11. org_rub
Export step 12. fil_rub
Export step 13. rub3
Export step 14. rub2
Export step 15. rub1


|     | ncalls | tottime | percall | cumtime | percall | filename:lineno(function)|
|-----|--------|---------|---------|---------|---------|--------------------------|
|Было | 240 |   8.280 |   0.035 |   8.570  | 0.036 |  dd.py:172(dexor_table)|
|Стало| 240  | 0.255  |  0.001  |  0.440  |  0.002 | dd_optimised.py:172(dexor_table)|